Include took time in DocWriteResponse

camilojd opened the pull request:
Closes #16641

jasontedor commented on core/src/main/java/org/elasticsearch/action/delete/TransportDeleteAction.java:
```diff
@@ -145,7 +145,8 @@ protected DeleteResponse newResponseInstance() {
         return new WriteResult<>(
-            new DeleteResponse(indexShard.shardId(), request.type(), request.id(), delete.version(), delete.found()),
+            new DeleteResponse(indexShard.shardId(), request.type(), request.id(), delete.version(), delete.found(),
+                               Math.max(1L, delete.endTime() - delete.startTime())),
```
The `Math.max` calls here and throughout are unnecessary; `System#nanoTime` is relative and will not go backwards.

jasontedor commented:
Can you please add tests? This is adding new functionality without tests, and we've had enough time and unit conversion bugs.

camilojd commented:
> But can you please add tests? This is adding new functionality without tests, and we've had enough time and unit conversion bugs.

Hi @jasontedor!

I addressed your review comment and added some asserts to code that already exercises `DocWriteResponse` and subclasses, to ensure `getTookInMillis() >= 0`.

Maybe it's necessary to write tests that ensure `getTookInMillis()` is effectively above zero for some predictable costly indexing operations?

jasontedor commented:
> I addressed your review comment and added some asserts to code that already exercises

@camilojd Did you squash and force push? It's easier for the review process if you just let the commits accumulate (unless a reviewer explicitly asks for a rebase/squash) :smile:

camilojd commented:
> @camilojd Did you squash and force push? It's easier for the review process if you just let the commits accumulate (unless a reviewer explicitly asks for a rebase/squash).

Yes, I did that. Sorry! Will take it into account for the next time :relieved:

jasontedor commented:
> to ensure `getTookInMillis() >= 0`

@camilojd Unfortunately this does not protect against any of the bugs that I'm worried about; there are many ways to change the code that will still satisfy this condition but be wrong.

> Maybe it's necessary to write tests that verify getTookInMillis() is effectively above zero for some predictable costly indexing operations?

Yeah. We likely need at least one test that tests the timing a controlled fashion, and one test that tests the timing with a real clock.

jasontedor commented:
> Yes, I did that. Sorry! Will take it into account for the next time :relieved:

@camilojd No worries, the basic rule is do not rewrite public history. For this PR it was relatively simple to go back and reread the entire diff, but you can imagine how this would be a problem on much larger change set that requires lots of back-and-forth.

olcbean commented:
@jasontedor 
I am currently looking at #23131 and I am wondering what tests should I write for it. Looking into the code base, I did not see any tests for the search and bulk 'took' time (maybe I missed something?). Hence I searched for hints in any related issues and saw your comment

> Yeah. We likely need at least one test that tests the timing a controlled fashion, and one test that tests the timing with a real clock.

As I am pretty new to ES, could you please elaborate what tests you have in mind as timing tests are really tricky (and rather unreliable).

jasontedor commented:
There is a test for bulk took time in `TransportBulkActionTookTests`.

A test with a controlled clock could be to ensure that the took calculations are correct. A test with a real clock could be to ensure that at least the took calculations produce sensible numbers.

olcbean commented:
@jasontedor 
Thank you for the pointer. I will have a look at it.

One more question : why is the bulk time measured in nanos while the search time is in millis? 
I myself do not see an obvious benefit of handling the time in nanos (especially when the return value is in millis). But nevertheless shouldn't we consider to unify the time measurement approach for both / all actions?

nik9000 commented:
> bulk time measured in nanos while the search time is in millis?

I haven't dug into it but if I had to guess it is because `System.nanoTime` is a relative clock while `System.currentTimeMillis` is an absolute clock. If you want to measure took time then `nanoTime` is better because it won't be affected by clock skew caused by systems like NTP.

jasontedor commented:
They are both reported in millis, but we absolutely should be using a high-precision *relative* clock for both search and bulk. I will fix this for search, bulk is fine.

jasontedor commented:
> They are both reported in millis, but we absolutely should be using a high-precision *relative* clock for both search and bulk. I will fix this for search, bulk is fine.

I opened #23662.
